sputnikdao2/src/proposals.rs

```diff
@@ -559,4 +565,30 @@ impl Contract {
             log!("Memo: {}", memo);
         }
     }
+
+    /// Receiving callback after the proposal has been finalized.
+    /// If successful, returns bond money to the proposal originator.
+    /// If the proposal execution failed (funds didn't transfer or function call failure),
+    /// move proposal to "Failed" state.
+    #[private]
+    pub fn on_proposal_callback(&mut self, proposal_id: u64) -> PromiseOrValue<()> {
+        let mut proposal: Proposal = self
+            .proposals
+            .get(&proposal_id)
+            .expect("ERR_NO_PROPOSAL")
+            .into();
+        assert_eq!(
+            env::promise_results_count(),
+            1,
+            "ERR_UNEXPECTED_CALLBACK_PROMISES"
+        );
+        let result = match env::promise_result(0) {
```

Comment:
Also, as a note we have great ergonomics around handling callbacks with results now. You can see an example here:
https://www.near-sdk.io/cross-contract/callbacks#whitelist-example
where we have essentially identical callbacks here:
https://github.com/mikedotexe/cross-contract-view/blob/06bdb3222622c824b9f2fe0a53536e6914435580/src/lib.rs#L56-L75

Except the second one uses `#[callback]` and saves quite a few lines.
cc @constantindogaru as an FYI 🎗️

Reply:
and actually looks like there might be further improvements according to the changelog:
https://github.com/near/near-sdk-rs/blob/master/CHANGELOG.md#400-pre3-10-12-2021

Reply:
Created https://github.com/near-daos/sputnik-dao-contract/issues/86 to address all of this.